### PR TITLE
Implement SO_REUSEPORT support

### PR DESCRIPTION
**Pull request by GlassOfWhiskey**
This PR adds support for the SO_REUSEPORT option in Gramine.
This PR is related to issues #238,  #361 and #763.

A regression test has been added to the `tcp_ipv6_v6only.c` file. Please let me know if other tests are needed.

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: 0 of 10 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @GlassOfWhiskey)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGLH6V8EKz7VNV0od_S:-NGLH6V8EKz7VNV0od_T:b-drvosf) (no related file):*
@glassofwhiskey Thanks for the PR! I quickly looked at it, and from a high level it looks good to me.

I would like to hear the opinions from @boryspoplawski and @mkow on this feature. An important historical context is probably this (also see linked GitHub issues in the PR description): https://github.com/gramineproject/gramine/issues/238#issuecomment-1127845795

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGLHZZO6EXFsOj0uwyL:-NGLHZZO6EXFsOj0uwyM:b-ip5ptx) (no related file):*
@glassofwhiskey May I kindly ask on which application/workload you required this functionality? Why did you decide to add it? How did it help (application refused to run without it, or was it purely a perf optimization)?

___

**Comment by GlassOfWhiskey** [comment]
> May I kindly ask on which application/workload you required this functionality? Why did you decide to add it? How did it help (application refused to run without it, or was it purely a perf optimization)?

Me and my colleagues at University of Torino are preparing a paper together with Intel in which we measure performance overhead of Federated Learning (FL) with TEE (and in particular Intel SGX).
We are using [OpenFL](https://github.com/intel/openfl), a FL framework similar to the [FLOWER](https://github.com/adap/flower) mentioned in #361. Under the hood, [gRPC](https://grpc.io/) deals with sockets and tries to set the `SO_REUSEPORT` option as a perf optimization, but it is not a strict requirement.
However, I was trying to remove all the "artificial" performance limitations, including `SO_REUSEPORT`. Since I saw that other people in the Gramine community asked for it in the past I decided to proceed with a PR.

> An important historical context is probably this (also see linked GitHub issues in the PR description): https://github.com/gramineproject/gramine/issues/238#issuecomment-1127845795

Here the comment mentions Windows. However, I didn't see any explicit adaptation layer for Windows. Did I miss it? In any case, pure Windows sockets can handle both Linux `SO_REUSEADDR` and Linux `SO_REUSEPORT` with different (and not obvious) combinations of their `SO_REUSEADDR` (which corresponds to Linux `SO_REUSEPORT`) and `SO_EXCLUSIVEADDRUSE` (see [here](https://learn.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse#enhanced-socket-security)).
Concerning instead the WSL, AFAIK it stil doesn't implement the `SO_REUSEPORT` explicitly (see https://github.com/microsoft/WSL/issues/1419). However, I am not a WSL dragon so I don't know if they already perform some adaptation between Linux and Windows versions of `SO_REUSEADDR` in their internals.

**Review by dimakuv — COMMENTED** [review]
@glassofwhiskey Thanks for the detailed reply! 

I'd be in favor of merging this PR.

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031#-:-NGLjGpLCbLSQdKVS56Y:bygm8xu)* status: 0 of 10 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @GlassOfWhiskey)

**Comment by GlassOfWhiskey** [comment]
> [-- commits line 4 at r1](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGQ_-ME1y7pOlCwui8Q):
Is this really the sign-off line you'd like to have, @GlassOfWhiskey ? Typically we have proper names and proper (work- or study-related) emails. But it's your decision of course.

I just force-pushed the commit with real name and work email in the `--signoff` line, as suggested. Next commits will be directly signed off with those credentials.

**Review by dimakuv — COMMENTED** [review]
> Reviewed 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @GlassOfWhiskey)

___
*[`libos/include/libos_handle.h` line 80 at r1](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGQ_bT285zxuvGLiuH3:-NG_ldPNB4pYCKM1Fzm7:bkx0vz3) ([raw file](https://github.com/gramineproject/gramine/blob/2ab7933f7f70241f7d46a7672e61c10e8a7c3f6d/libos/include/libos_handle.h#L80)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

What about accessing `reuseaddr`, `broadcast` and now this new `reuseport` fields? Do we need to grab a lock here? Question to @boryspoplawski 
</blockquote></details>

This question is unrelated to this particular PR, so let me unblock this discussion. Still would like to know the answer.

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @GlassOfWhiskey)

**Comment by dimakuv** [comment]
Jenkins, test this please

**Comment by GlassOfWhiskey** [comment]
Starting tomorrow, I'll be away for one week for a conference. There is anything else I have to do with this PR before leaving?

**Review by dimakuv — COMMENTED** [review]
@glassofwhiskey Nothing to be done on your side. We're waiting for one more review before proceeding with this PR (we need at least two reviews for PRs).

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031#-:-NGuipti9NTApwoSX7p_:bbgfudc)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @GlassOfWhiskey)

**Review by boryspoplawski — DISMISSED** [review]
> Reviewed 1 of 10 files at r1, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: all files reviewed, 4 unresolved discussions (waiting on @dimakuv and @GlassOfWhiskey)

___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/gramineproject/gramine/1031#-NHgC8MO1st18qjZaxv8):*
We need to not forget to squash this commit
___
*[`libos/include/libos_handle.h` line 80 at r1](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGQ_bT285zxuvGLiuH3:-NHgCvP85Dqj1Q9zDgTp:budjfkk) ([raw file](https://github.com/gramineproject/gramine/blob/2ab7933f7f70241f7d46a7672e61c10e8a7c3f6d/libos/include/libos_handle.h#L80)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This question is unrelated to this particular PR, so let me unblock this discussion. Still would like to know the answer.
</blockquote></details>

Yes, all 3 should be protected by `lock` of this `sock` struct.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski and @GlassOfWhiskey)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGLH6V8EKz7VNV0od_S:-NHhf_4s3PYNOnoi2-HW:bcrgekl) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@glassofwhiskey Thanks for the PR! I quickly looked at it, and from a high level it looks good to me.

I would like to hear the opinions from @boryspoplawski and @mkow on this feature. An important historical context is probably this (also see linked GitHub issues in the PR description): https://github.com/gramineproject/gramine/issues/238#issuecomment-1127845795
</blockquote></details>

Looks like Borys is happy with this PR, unblocking.

___
___
*[`libos/include/libos_handle.h` line 80 at r1](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGQ_bT285zxuvGLiuH3:-NHhffCPE3fe0Of5_kMm:b-e4o8h2) ([raw file](https://github.com/gramineproject/gramine/blob/2ab7933f7f70241f7d46a7672e61c10e8a7c3f6d/libos/include/libos_handle.h#L80)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Yes, all 3 should be protected by `lock` of this `sock` struct.
</blockquote></details>

Let me block here, and I'll add this info during final rebase (doesn't deserve to be in a separate PR)

**Review by mkow — DISMISSED** [review]
> Reviewed 1 of 10 files at r1, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: all files reviewed, 3 unresolved discussions, not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski and @GlassOfWhiskey)

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGQ_C2F66Ve0gE-F-3C):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Should have `[LibOS,PAL]` prefix. We can add it during final rebase, no need for you to do anything here.
</blockquote></details>

Done
___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/gramineproject/gramine/1031#-NHgC8MO1st18qjZaxv8):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

We need to not forget to squash this commit
</blockquote></details>

Done
___
*[`libos/include/libos_handle.h` line 80 at r1](https://reviewable.io/reviews/gramineproject/gramine/1031#-NGQ_bT285zxuvGLiuH3:-NHsxd6927DyGJKK0iJc:bb6un1j) ([raw file](https://github.com/gramineproject/gramine/blob/2ab7933f7f70241f7d46a7672e61c10e8a7c3f6d/libos/include/libos_handle.h#L80)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Let me block here, and I'll add this info during final rebase (doesn't deserve to be in a separate PR)
</blockquote></details>

Done

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by boryspoplawski — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1031)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
